x-pack/plugins/security_solution/common/detection_engine/rule_response_actions/schemas/osquery.ts

```diff
@@ -7,27 +7,46 @@
 
 import * as t from 'io-ts';
 
+const ecsMapping = t.record(
+  t.string,
+  t.partial({
+    field: t.string,
+    value: t.union([t.string, t.array(t.string)]),
+  })
+);
+
+export const arrayQueries = t.array(
+  t.type({
+    id: t.string,
+    query: t.string,
+    ecs_mapping: t.union([ecsMapping, t.undefined]),
+    version: t.union([t.string, t.undefined]),
+    platform: t.union([t.string, t.undefined]),
+  })
+);
+
 export const OsqueryParams = t.intersection([
   t.type({
     id: t.string,
   }),
   t.partial({
-    query: t.union([t.string, t.undefined]),
-    ecs_mapping: t.record(t.string, t.record(t.string, t.any)),
-    queries: t.array(
-      t.intersection([
-        t.type({
-          id: t.string,
-          query: t.string,
-        }),
-        t.partial({
-          ecs_mapping: t.record(t.string, t.record(t.string, t.any)),
-          platform: t.union([t.string, t.undefined]),
-          interval: t.union([t.number, t.undefined]),
-        }),
-      ])
-    ),
-    packId: t.union([t.string, t.undefined]),
-    savedQueryId: t.union([t.string, t.undefined]),
+    query: t.string,
+    ecs_mapping: ecsMapping,
+    queries: arrayQueries,
+    pack_id: t.string,
+    saved_query_id: t.string,
+  }),
+]);
+
+export const OsqueryParamsCamelCase = t.intersection([
+  t.type({
+    id: t.string,
+  }),
+  t.partial({
```

Comment:
I would recommend using the pattern `t.union([<your type>, t.undefined])` to make fields optional instead of `t.partial`. `t.partial` makes it easy to accidentally drop fields in the conversion between HTTP and internal schema. You can try it out in `transformRuleToAlertResponseAction` by removing individual fields, e.g. `packId`, from the return value, there's no type error - but with 
```
t.type({
  id: t.string,
  packId: t.union([t.string, t.undefined]),
}),
```
the compiler will generate an error if `packId` is missing from the conversion.

The same idea applies to OsqueryParams above. It can make creating mock data a bit more tedious, since fields must be explicitly set to undefined, but it helps to eliminate subtle bugs in the schema logic.

Reply:
Alright, makes sense, will pass unions with undefined :)